### PR TITLE
DT-961 index service

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Implement mark build index complete and cancel build index services

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt [hunk 1]
```diff
@@ -32,15 +32,43 @@ class IndexService(
...
-  fun markIndexingComplete() = log.info("Received request to mark indexing complete")
+  fun markIndexingComplete(): Either<MarkBuildCompleteError, IndexStatus> {
+    val indexStatus = indexStatusService.getOrCreateCurrentIndexStatus()
+    if (indexStatus.state != IndexState.BUILDING) {
+      return MarkBuildCompleteError.BuildNotInProgress(indexStatus).left()
+    }
+
+    indexStatusService.markBuildComplete()
+    indexQueueService.clearAllMessages()
+    log.info("Index ${indexStatus.currentIndex.otherIndex()} marked as complete, ${indexStatus.currentIndex} is now current")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt:43]: Interpolation in log statements is best avoided (IMO), as I understand it the string will evaluate even if log.info is disabled.  Apparently can be avoided if used a Lamda construction or just the original/standard way

https://logging.apache.org/log4j/kotlin/#usage

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt [hunk 2]
```diff
@@ -32,15 +32,43 @@ class IndexService(
...
   fun indexOffender(crn: String) = offenderSynchroniserService.synchroniseOffender(crn)
 }
 
 
-@Suppress("UNUSED_PARAMETER")
 sealed class BuildIndexError(val message: String) {
   data class BuildAlreadyInProgress(val indexStatus: IndexStatus): BuildIndexError("The build for ${indexStatus.currentIndex} is already ${indexStatus.state} (started at ${indexStatus.startIndexTime})")
+}
+
+sealed class MarkBuildCompleteError(val message: String) {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt:68]: If we end up only having one subclass will you remove the subclass and simplify the handling resource code? eg get rid of the `when`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/resource/IndexResource.kt [hunk 3]
```diff
@@ -23,19 +25,32 @@ class IndexResource(private val indexService: IndexService) {
...
+          .getOrHandle { error ->
+            when (error) {
+              is BuildIndexError.BuildAlreadyInProgress -> throw ResponseStatusException(HttpStatus.CONFLICT, error.message)
+            }
+          }
 
   @PutMapping("/mark-complete")
   @PreAuthorize("hasRole('PROBATION_INDEX')")
-  fun markComplete() = indexService.markIndexingComplete()
+  fun markComplete() =
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/resource/IndexResource.kt:37]: Just a general question.  Is there a plan to add Swagger docs?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexServiceTest.kt [hunk 4]
```diff
@@ -34,7 +32,7 @@ class IndexServiceTest {
       val result = indexService.buildIndex()
 
       verify(indexStatusService).getOrCreateCurrentIndexStatus()
-      assertThat(result).isEqualTo(BuildIndexError.BuildAlreadyInProgress(expectedIndexStatus).left())
+      assertThat(result.getOrHandle { it }).isEqualTo(BuildIndexError.BuildAlreadyInProgress(expectedIndexStatus))
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexServiceTest.kt:35]: I'm curious, what type does `result.getOrHandle { it }` return?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt [hunk 5]
```diff
@@ -32,15 +32,43 @@ class IndexService(
...
+  }
+
+  fun cancelIndexing(): Either<CancelBuildIndexError, IndexStatus> {
+    val indexStatus = indexStatusService.getOrCreateCurrentIndexStatus()
+    if (indexStatus.state != IndexState.BUILDING) {
+      return CancelBuildIndexError.BuildNotInProgress(indexStatus).left()
+    }
+
+    indexStatusService.markBuildCancelled()
+    log.info("Index ${indexStatus.currentIndex.otherIndex()} marked as cancelled, ${indexStatus.currentIndex} is now current")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt:55]: Same as other logging comment above.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt (hunk 2 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt:68]: A major benefit of this approach is that when I look at the API I see all possible errors at the API level (vs trawling through code looking for exceptions).  I don't want to hide those errors - even if there's only one.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexServiceTest.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexServiceTest.kt:35]: Whichever side of the Either is active (if that's the right word).  e.g. the left side if there is an error or the right side if there isn't.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt (hunk 1 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/service/IndexService.kt:43]: Cool - I'll switch to lambdas in the next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/resource/IndexResource.kt (hunk 3 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/indexer/resource/IndexResource.kt:37]: Next PR :smile:

**Pull request merged**